LoRaEngine/LoraKeysManagerFacade/ConcentratorFirmwareFunction.cs

```diff
@@ -0,0 +1,118 @@
+            }
+
+            return await RunFetchConcentratorFirmware(req, cancellationToken);
+        }
+
+        internal async Task<IActionResult> RunFetchConcentratorFirmware(HttpRequest req, CancellationToken cancellationToken)
+        {
+            if (!StationEui.TryParse((string)req.Query["StationEui"], out var stationEui))
+            {
+                this.logger.LogError("StationEui missing in request or invalid");
+                return new BadRequestObjectResult("StationEui missing in request or invalid");
+            }
+
+            var twin = await this.registryManager.GetTwinAsync(stationEui.ToString("N", CultureInfo.InvariantCulture), cancellationToken);
+            if (twin != null)
+            {
+                this.logger.LogDebug("Retrieving firmware url for '{StationEui}'.", stationEui);
+                try
+                {
+                    if (!twin.Properties.Desired.TryReadJsonBlock(CupsPropertyName, out var cupsProperty))
+                        throw new ArgumentOutOfRangeException(CupsPropertyName, "Failed to read CUPS config");
+
+                    var fwUrl = JObject.Parse(cupsProperty)[CupsFwUrlPropertyName].ToString();
+                    var stream = await GetBlobStreamAsync(fwUrl, cancellationToken);
+                    return new FileStreamResult(stream, MediaTypeNames.Text.Plain);
+                }
+                catch (Exception ex) when (ex is ArgumentOutOfRangeException or JsonReaderException or NullReferenceException)
+                {
+                    this.logger.LogError(ex, "Failed to parse firmware upgrade url from the '{PropertyName}' desired property.", CupsPropertyName);
+                    return new UnprocessableEntityResult();
+                }
+                catch (RequestFailedException ex)
+                {
+                    this.logger.LogError(ex, "Failed to download firmware from storage.");
+                    return new ObjectResult("Failed to download firmware")
+                    {
+                        StatusCode = (int)HttpStatusCode.InternalServerError
+                    };
+                }
+            }
+            else
+            {
+                this.logger.LogInformation($"Searching for {stationEui} returned 0 devices");
+                return new NotFoundResult();
+            }
+        }
+
+        private async Task<MemoryStream> GetBlobStreamAsync(string blobUrl, CancellationToken cancellationToken)
+        {
+            var blobServiceClient = this.azureClientFactory.CreateClient(FacadeStartup.WebJobsStorageClientName);
+            var blobUri = new BlobUriBuilder(new Uri(blobUrl));
+            var stream = new MemoryStream();
+            _ = await blobServiceClient.GetBlobContainerClient(blobUri.BlobContainerName)
+                                       .GetBlobClient(blobUri.BlobName)
+                                       .DownloadToAsync(stream, cancellationToken);
```

Comment:
We should not use a memory stream and download the firmware into function's memory.
Please consider usage of DownloadStreamingAsync which will expose the underlying network stream, without loading anything in memory

Reply:
Addressed in 10e041b39d59a3e09c43eaf99c4246ae7e4972c7, please reopen if you still have some comments regarding this